internal/replication/blockreplicator_3node_test.go

```diff
@@ -0,0 +1,313 @@
+		if err != nil && err.LeaderID == 0 {
+			return true
+		}
+		return false
+	}
+	assert.Eventually(t, noLeaderCond, 30*time.Second, 100*time.Millisecond)
+
+	err = env.nodes[lastIndex].Close()
+	require.NoError(t, err)
+}
+
+// Scenario:
+// - Start 1 node, no leader
+// - Start 2nd node, wait for leader,
+// - Start 3rd, wait for consistent leader on all
+func TestBlockReplicator_3Node_StartStepClose(t *testing.T) {
+	env := createClusterEnv(t, "info", 3)
+	require.Equal(t, 3, len(env.nodes))
+	defer os.RemoveAll(env.testDir)
+
+	err := env.nodes[0].Start()
+	require.NoError(t, err)
+	noLeaderCond := func() bool {
+		err := env.nodes[0].blockReplicator.IsLeader()
+		if err != nil && err.LeaderID == 0 {
+			return true
+		}
+		return false
+	}
+	assert.Eventually(t, noLeaderCond, 30*time.Second, 100*time.Millisecond)
+
+	err = env.nodes[1].Start()
+	require.NoError(t, err)
+	isLeaderCond1 := func() bool {
+		return env.AgreedLeaderIndex(0, 1) >= 0
+	}
+	assert.Eventually(t, isLeaderCond1, 30*time.Second, 100*time.Millisecond)
+
+	err = env.nodes[2].Start()
+	require.NoError(t, err)
+	isLeaderCond2 := func() bool {
+		return env.AgreedLeaderIndex() >= 0
+	}
+	assert.Eventually(t, isLeaderCond2, 30*time.Second, 100*time.Millisecond)
+
+	//close all
+	for _, node := range env.nodes {
+		err := node.Close()
+		require.NoError(t, err)
+	}
+}
+
+// Scenario:
+// - Start 3 nodes together, wait for leader,
+// - Stop the leader, wait for new leader,
+// - Restart the node, wait for consistent leader on all
+// - Stop a follower, wait for new leader,
+// - Restart the node, wait for consistent leader on all
+func TestBlockReplicator_3Node_Restart(t *testing.T) {
+	env := createClusterEnv(t, "info", 3)
+	require.Equal(t, 3, len(env.nodes))
+	defer os.RemoveAll(env.testDir)
+
+	// start 3 at once
+	for _, node := range env.nodes {
+		err := node.Start()
+		require.NoError(t, err)
+	}
+
+	// wait for an agreed leader
+	sameLeaderCond := func() bool {
+		return env.AgreedLeaderIndex() >= 0
+	}
+	assert.Eventually(t, sameLeaderCond, 30*time.Second, 100*time.Millisecond)
+
+	//close the leader, wait for some node to become a new leader
+	leaderIndex1 := env.FindLeaderIndex()
+	require.True(t, leaderIndex1 >= 0)
+	t.Logf("Stopping leader, index: %d", leaderIndex1)
+	err := env.nodes[leaderIndex1].Close()
+	require.NoError(t, err)
+	isLeaderCond := func() bool {
+		return env.FindLeaderIndex() >= 0
+	}
+	assert.Eventually(t, isLeaderCond, 30*time.Second, 100*time.Millisecond)
+
+	//restart the old leader
+	t.Logf("Restarting node, index: %d", leaderIndex1)
+	err = env.nodes[leaderIndex1].Restart()
+	require.NoError(t, err)
+	assert.Eventually(t, sameLeaderCond, 30*time.Second, 100*time.Millisecond)
+
+	leaderIndex2 := env.FindLeaderIndex()
+	require.True(t, leaderIndex2 >= 0)
+	followerIndex := (leaderIndex1 + 1) % 3
+	t.Logf("Stopping follower, index: %d", followerIndex)
+	err = env.nodes[followerIndex].Close()
```

Comment:
How does `(leaderIndex1 + 1) %3` provide an index of a follower? Was it supposed to be `leaderIndex2` instead of `leaderIndex1`?

Reply:
good catch!